### PR TITLE
Validate device protocol configs based on webhook

### LINKED ISSUE #826: devicecrd: Reject protocol configs for protocols other than OPC UA , Modbus and Bluetooth.
**What would you like to be added**:
While creating a device instance, protocol configs are optional, but if provided, they can be one of [OPC UA, Modbus, and Bluetooth]. Other protocol configs are not supported.

**Why is this needed**:
Validation of CRDs is needed for ensuring valid metadata is persisted for use by mappers/platform.

### PR DESCRIPTION
**Pull request by chendave**
Signed-off-by: Dave Chen <dave.chen@arm.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #826

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by chendave** [comment]
cc @fisherxu @rohitsardesai83 @subpathdev @sids-b  @kadisi

**Comment by chendave** [comment]
cc @kevin-wangzefeng

**Comment by fisherxu** [comment]
@chendave  Need rebase, we can add a file named admin_device to deal with the device now, how do you think? :)

**Comment by chendave** [comment]
@fisherxu , sure, will do.

**Comment by kevin-wangzefeng** [comment]
We should not reject protocols that is not predefined in-tree, which will prevent people adding their own protocol/mapper implementations.
For protocols already defined in-tree, we can check if all the fields are filled correctly.

**Comment by chendave** [comment]
> We should not reject protocols that is not predefined in-tree, which will prevent people adding their own protocol/mapper implementations.

Agreed, so we might need to update the bug there. @rohitsardesai83 

> For protocols already defined in-tree, we can check if all the fields are filled correctly.

we don't need to do that, this case has been covered by the validation rules defined in the yaml.
so, all we need to do here is make sure only one protocol is defined for each device as what https://github.com/kubeedge/kubeedge/issues/826 said.

File: cloud/pkg/admissioncontroller/admission.go [hunk 1]
```diff
@@ -195,3 +231,57 @@ func registerValidateWebhook(client admissionregistrationv1beta1client.Validatin
...
+		responseAdmissionReview.Response = admit(requestedAdmissionReview)
+	}
+
+	// Return the same UID
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+	klog.Infof("sending response: %v", responseAdmissionReview.Response)
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		klog.Fatalf("cannot marshal to a valid reponse %v", err)
```
Comment (rohitsardesai83) [inline cloud/pkg/admissioncontroller/admission.go]: reponse  -> response

Comment (rohitsardesai83) [reply cloud/pkg/admissioncontroller/admission.go]: applies at all such places

Comment (rohitsardesai83) [reply cloud/pkg/admissioncontroller/admission.go]: what do we return here ?

Comment (chendave) [reply cloud/pkg/admissioncontroller/admission.go]: thanks, addressed!

**Comment by rohitsardesai83** [comment]
@chendave , can you please fix the lint issue reported by the CI.

File: cloud/pkg/admissioncontroller/admission.go [hunk 2]
```diff
@@ -195,3 +231,58 @@ func registerValidateWebhook(client admissionregistrationv1beta1client.Validatin
...
+// admitFunc is the type we use for all of our validators and mutators
+type admitFunc func(admissionv1beta1.AdmissionReview) *admissionv1beta1.AdmissionResponse
+
+func serve(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
```
Comment (sids-b) [inline cloud/pkg/admissioncontroller/admission.go]: should we return if ``` r.Body == nil ```

File: cloud/pkg/admissioncontroller/admission.go [hunk 3]
```diff
@@ -195,3 +231,58 @@ func registerValidateWebhook(client admissionregistrationv1beta1client.Validatin
...
+
+	// The AdmissionReview that was sent to the webhook
+	requestedAdmissionReview := admissionv1beta1.AdmissionReview{}
+
+	// The AdmissionReview that will be returned
+	responseAdmissionReview := admissionv1beta1.AdmissionReview{}
+
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		klog.Fatalf("decode failed with error: %v", err)
```
Comment (sids-b) [inline cloud/pkg/admissioncontroller/admission.go]: Just a question. Do we need fatal log ? Fatal log will indicate stopping the service...

File: cloud/pkg/admissioncontroller/admission.go [hunk 4]
```diff
@@ -195,3 +231,58 @@ func registerValidateWebhook(client admissionregistrationv1beta1client.Validatin
...
+	}
+	responseAdmissionReview.Response = admit(requestedAdmissionReview)
+
+	// Return the same UID
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+	klog.Infof("sending response: %v", responseAdmissionReview.Response)
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		klog.Fatalf("cannot marshal to a valid response %v", err)
```
Comment (sids-b) [inline cloud/pkg/admissioncontroller/admission.go]: nit: please start logs with caps.
Applicable to all similar places

File: cloud/pkg/admissioncontroller/admit_device.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+		mdSpec, ok := un.Object["spec"].(map[string]interface{})
+		if ok {
+			mdProto, ok := mdSpec["protocol"].(map[string]interface{})
+			if ok {
+				keys := reflect.ValueOf(mdProto).MapKeys()
+				//device should only have one protocol defined.
+				if len(keys) > 1 {
+					reviewResponse.Allowed = false
+					msg = msg + "multiple protocols found!"
```
Comment (sids-b) [inline cloud/pkg/admissioncontroller/admit_device.go]: nit: please add a blank space either before or after the appended message to make it more readable

File: cloud/pkg/admissioncontroller/admission.go (hunk 2 above)
Comment (chendave) [reply cloud/pkg/admissioncontroller/admission.go]: I think so, thanks! will address this soon.

File: cloud/pkg/admissioncontroller/admit_device.go (hunk 5 above)
Comment (chendave) [reply cloud/pkg/admissioncontroller/admit_device.go]: Good catch!

**Comment by sids-b** [comment]
Thanks @chendave for promptly addressing all the comments.

/approve

**Comment by chendave** [comment]
let's revisit this later if this is needed, close it for now.

**Pull request closed without merging**